simde/simde-arch.h

```diff
@@ -302,7 +314,7 @@
 #  if defined(__GFNI__)
 #    define SIMDE_ARCH_X86_GFNI 1
 #  endif
-#  if defined(__PCLMUL__)
+#  if defined(__PCLMUL__) && !defined(SIMDE_ARCH_E2K) /* E2K has inefficient implementation of PCLMUL */
```

Comment:
> Wow, slower than the portable implementation in SIMDe?

I actually don't know; I didn't measure it. LCC just produces "deprecation" warnings that state something like "this function shall not be used due to inefficient implementation". Maybe it would be nice to check actual speed. Or just rely on "even if native implementation is inefficient, it might be better than software" and suppress these warnings. Still I have no idea how to benchmark these two implementations using SIMDe.

>  simde-arch.h is intended to be usable outside of SIMDe

Ok, so I should use '__e2k__' instead of `SIMDE_ARCH_E2K` here, right?

Reply:
> Still I have no idea how to benchmark these two implementations using SIMDe.

Most of my performance testing for SIMDe is done using llvm-mca. We don't really have anything set up; I've used uarch-bench and google-benchmark in the past… the latter should be pretty straightforward to set up a new test for, and I may even have some old code somewhere. I'll try to get to that at some point in the near future.

> Ok, so I should use 'e2k' instead of SIMDE_ARCH_E2K here, right?

In simde-arch.h you should define `SIMDE_ARCH_E2K`, then change `#if defined(SIMDE_ARCH_X86_PCLMUL)` to `#if defined(SIMDE_ARCH_X86_PCLMUL) && !defined(SIMDE_ARCH_E2K)` in simde-features.h when it's checking whether to define `SIMDE_X86_PCLMUL_NATIVE`.  You may want to do the same thing for VPCLMULQDQ (if the CPU also says it is supported but is deprecated).

That said, if the e2k implementation is really faster (which seems likely), I think the right thing to do would be to just suppress the deprecation warning.  Assuming it's just a function with an `__attribute__((__deprecated__))` or similar, usually this is as simple as

```
HEDLEY_DIAGNOSTIC_PUSH
HEDLEY_DIAGNOSTIC_DISABLE_DEPRECATED
deprecated_call();
HEDLEY_DIAGNOSTIC_POP
```

Unfortunately (AFAICT) LCC doesn't support pushing/popping the warning stack, so this would really get mapped to just 

```
#pragma diag_suppress 1215,1444
deprecated_call();
```

Which isn't ideal since the user may want those warnings enabled later.  Still, I don't see a better solution.

Looking at the include dir for lcc, maybe it would be possible to implement this using `__builtin_e2k_clmull` and/or `__builtin_e2k_clmulh` instead of calling the native x86 functions?

Reply:
Implemented in f4817b6.
There is `SIMDE_LCC_DISABLE_INEFFICIENT_FUNCTIONS` macro that can be defined at simde-arch.h:202 to use software implementations; otherwise, native ones are used, and deprecation warnings are suppresses in `xop.h` and `fma.h`.
Don't know if this is what we want (shouldn't these functions be disabled by default?), but IMO looks better than totally cancelling out these inefficient functions unconditionally.

Reply:
In my newest rebase on `wip/e2k`, I decided to just use the native functions, but wrap them around in `diag_suppress`/`diag_default` using macros like `SIMDE_LCC_DISABLE_DEPRECATED_WARNINGS` and `SIMDE_LCC_REVERT_DEPRECATED_WARNINGS`. This seems to be the best approach I could think of.

Reply:
Looks like we agreed that it's the right way, so looks like it's resolved too.